components/camel-datasonnet/src/main/java/org/apache/camel/language/datasonnet/DatasonnetExpression.java

```diff
@@ -114,7 +114,11 @@ private Document<?> doEvaluate(Exchange exchange) {
         }
 
         Document<?> body;
-        if (exchange.getMessage().getBody() instanceof Document) {
+
+        if (exchange.getMessage().getBody() == null || "".equals(MessageHelper.extractBodyAsString(exchange.getMessage()))) {
```

Comment:
This logic should be moved last as extracting body as string would then always be performed and that is not good

Reply:
Oh, I see, but then how do I verify if the body is an empty string?

Reply:
Ah yeah do you need to do this even if there is a Document object and its empty? Because if you convert that to a string you do an xml dump which would not be empty, it will have <xml ...> something as output.

The extractBody ... method deals with stream caching so that is good. It may not be as much overkill, just that the message body can be InputStream, byte[], XML Document and many other things, and in Java everything can be converted to a String via Object.toString so you would essentailly always get some output.

Reply:
No, if it's an empty Document, it's a different story (a Document cannot really be empty). Datasonnet doesn't like empty inputs, and I'm really just trying to deal with the situation where message body is not used in mapping but Datasonnet expects it anyway. If an input is an XML document and the mapping expects something else, there should be a legitimate error thrown.
So I really just need to make sure that if it's null or an empty string, we force the empty string with java object mime type, to avoid an error.

Reply:
OK, I changed the logic slightly, but I still need to ensure that if it's null or an empty string, I force APPLICATION_JAVA with empty string to DS.